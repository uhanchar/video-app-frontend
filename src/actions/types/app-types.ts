export const LOAD_VIDEOS = '[APP] LOAD_VIDEOS';
