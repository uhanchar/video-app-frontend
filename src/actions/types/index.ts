export * from 'actions/types/app-types';
export * from 'actions/types/video-types';
