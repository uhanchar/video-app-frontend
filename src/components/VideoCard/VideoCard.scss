@import "src/constants";

.video-card {
  width: 100%;
  cursor: pointer;

  > .media {
    position: relative;
    padding-top: 56.25%;

    > .player {
      position: absolute;
      top: 0;
      left: 0;
      width: 100% !important;
      height: 100% !important;

      > video {
        background: $primary-color-black;
      }
    }
  }

  > .content {
    height: 1rem;

    > .description {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
  }
}
