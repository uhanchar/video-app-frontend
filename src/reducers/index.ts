import { combineReducers } from 'redux';
import AppReducer from 'reducers/app.reducer';

const rootReducer = combineReducers({
  app: AppReducer,
});

export default rootReducer;
