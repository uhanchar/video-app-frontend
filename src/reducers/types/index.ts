import { NullableType } from 'constants/data-types';

export interface IVideoItem {
  id: string;
  name: string;
  description: string;
  link: string;
  thumbnailLink: string;
  createdAt: Date;
}

export interface IAppState {
  videos: IVideoItem[];
}

export interface IVideoState {
  video: NullableType<IVideoItem>;
  isLoading: boolean;
  error: string;
}

export type RootState = {
  app: IAppState;
  video: IVideoState;
};
