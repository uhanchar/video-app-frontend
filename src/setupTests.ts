// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';

export const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  // @ts-ignore
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));
