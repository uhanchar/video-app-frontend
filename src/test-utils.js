import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { applyMiddleware, createStore } from 'redux';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import thunk from 'redux-thunk';

import rootReducer from './reducers';

const reducerInitialState = {};

function render(
  ui,
  {
    initialState = reducerInitialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...renderOptions
  } = {},
) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return (
      // eslint-disable-next-line react/jsx-filename-extension
      <Provider store={store}>
        <BrowserRouter>
          {children}
        </BrowserRouter>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';

export { render };
