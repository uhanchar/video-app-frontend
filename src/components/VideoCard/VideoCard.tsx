import React, { useState } from 'react';
import { Card, CardHeader, CardMedia, CardContent, Typography } from '@material-ui/core';

import 'components/VideoCard/VideoCard.scss';
import { IVideoItem } from 'reducers/types';
import { useHistory } from 'react-router-dom';
import { LocationPath } from 'constants/location-path.enum';
import ReactPlayer from 'react-player';
import { formatValidDate } from 'helpers/date-format.helper';

const VideoCard = (props: IVideoItem) => {
  const { name, description, createdAt, link, id } = props;
  const history = useHistory();
  const [ isPlaying, setIsPlaying ] = useState<boolean>(false);

  const onCardClick = () => {
    history.push(`${ LocationPath.Video }/${ id }`);
  };

  const onMouseVideoEnter = () => {
    setIsPlaying(true);
  };

  const onMouseVideoLeave = () => {
    setIsPlaying(false);
  };

  return (
    <Card className="video-card" onClick={onCardClick}>
      <CardHeader
        title={name}
        subheader={formatValidDate(createdAt)}
      />

      <CardMedia className="media" onMouseEnter={onMouseVideoEnter} onMouseLeave={onMouseVideoLeave}>
        <ReactPlayer url={`/${ link }`} className="player" loop playing={isPlaying} volume={0} />
      </CardMedia>

      <CardContent className="content">
        <Typography
          variant="body2"
          color="textSecondary"
          component="p"
          className="description"
        >
          {description}
        </Typography>
      </CardContent>
    </Card>
  );
};

export default VideoCard;
