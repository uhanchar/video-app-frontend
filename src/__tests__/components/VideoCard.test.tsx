import React from 'react';
import { fireEvent } from '@testing-library/react';

import { render as renderWithRedux } from 'test-utils';
import VideoCard from 'components/VideoCard/VideoCard';
import { IVideoItem } from 'reducers/types';
import { mockHistoryPush } from 'setupTests';
import { LocationPath } from 'constants/location-path.enum';

const videoItem: IVideoItem = {
  id: '39432ff1585cfb0ed1803ccea2929341',
  name: 'enem2',
  description: '',
  link: 'media/videos/39432ff1585cfb0ed1803ccea2929341',
  thumbnailLink: 'media/thumbnails/39432ff1585cfb0ed1803ccea2929341.png',
  createdAt: new Date(),
};

describe('test for VideoCard component', () => {
  it('should render VideoCard component', () => {
    const { asFragment } = renderWithRedux(<VideoCard {...videoItem} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should redirect to Video component on Card component click', () => {
    const { getByTestId } = renderWithRedux(<VideoCard {...videoItem} />);

    fireEvent.click(getByTestId('video-card'));

    expect(mockHistoryPush).toHaveBeenCalledWith(`${ LocationPath.Video }/${ videoItem.id }`);
  });
});
