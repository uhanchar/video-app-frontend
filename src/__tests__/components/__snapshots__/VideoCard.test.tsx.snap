// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for VideoCard component should render VideoCard component 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root video-card MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="video-card"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          enem2
        </span>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          less than a minute ago
        </span>
      </div>
    </div>
    <div
      class="MuiCardMedia-root media"
      data-testid="video-card-media"
    >
      <div
        class="player"
        data-testid="video-card-player"
        style="width: 640px; height: 360px;"
      />
    </div>
    <div
      class="MuiCardContent-root content"
    >
      <p
        class="MuiTypography-root description MuiTypography-body2 MuiTypography-colorTextSecondary"
      />
    </div>
  </div>
</DocumentFragment>
`;
